src/libDirectoryService/DirectoryService.cpp

```diff
@@ -507,3 +507,82 @@ std::string DirectoryService::GetActionString(Action action) const
         ? "Unknown"
         : ActionStrings.at(action);
 }
+
+uint8_t
+DirectoryService::CalculateNewDifficulty(const uint8_t& currentDifficulty)
+{
+    const uint8_t MAX_ADJUST_STEP = 2;
+    int requiredNodes = 0, powSubmissions = 0;
+    {
+        lock_guard<mutex> g(m_mutexAllPOW);
+        powSubmissions = static_cast<int>(m_allPoWs.size());
+
+        for (const auto& shard : m_shards)
+        {
+            requiredNodes += shard.size();
+        }
+    }
+
+    LOG_GENERAL(INFO,
+                "requiredNodes " << requiredNodes << ", powSubmissions "
+                                 << powSubmissions);
+
+    int adjustment = 0;
+    if (requiredNodes > 0 && requiredNodes != powSubmissions)
+    {
+        int submissionsDiff = powSubmissions - requiredNodes;
+
+        // To make the adjustment work on small network.
+        int adjustThreshold = requiredNodes * 0.12f;
```

Comment:
consider making magic number into constants in constants file

Reply:
OK, i will put this magic number into constants file.